crates/ra_hir/src/expr/validation.rs

```diff
@@ -87,4 +100,43 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
             })
         }
     }
+
+    fn validate_results_in_tail_expr(&mut self, id: ExprId, db: &impl HirDatabase) {
+        let expr_ty = &self.infer[id];
+        let func_ty = self.func.ty(db);
+        let func_sig = func_ty.callable_sig(db).unwrap();
+        let ret = func_sig.ret();
+        let ret = match ret {
+            Ty::Apply(t) => t,
+            _ => return,
+        };
+        let ret_enum = match ret.ctor {
+            TypeCtor::Adt(AdtDef::Enum(e)) => e,
+            _ => return,
+        };
+        let enum_name = ret_enum.name(db);
+        if enum_name.is_none() || enum_name.unwrap().to_string() != "Result" {
```

Comment:
It would be a bit more robust to resolve the path `std::result::Result` and compare the IDs, or maybe at least check that the type has a variant `Ok`.

Reply:
Yes, this bit was deliberately a WIP hack just to get something working. I'll try to make it more robust (in a way that still lets my tests-without-std-types pass!)

Reply:
You can look at how we're mocking `std::future::Future` trait from stdlib in tests: 

https://github.com/rust-analyzer/rust-analyzer/blob/4912cc35afe96b1506e5cda1be34d3e2418bc702/crates/ra_ide_api/src/completion/complete_dot.rs#L440-L443

And how to get it by its path:

https://github.com/rust-analyzer/rust-analyzer/blob/4912cc35afe96b1506e5cda1be34d3e2418bc702/crates/ra_hir/src/source_binder.rs#L419-L432

and do similar for `std::result::Result` enum instead.

Reply:
Perfect, thanks!